File: Data Structures & Algorithms/house-robber/submission-3.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 0:
            return 0
        
        nums[len(nums)-2] = max(nums[len(nums)-1], nums[len(nums)-2])
        for i in range(len(nums)-1-2, -1, -1):
            nums[i] = max(nums[i] + nums[i+2], nums[i+1])
        return max(nums[0], nums[1])

File: Data Structures & Algorithms/maximum-product-subarray/submission-3.java
class Solution {
    public int maxProduct(int[] nums) {
        int max = 1;
        int min = 1;
        int res = nums[0];
        for(int num: nums) {
            int temp = max*num;
            max = Math.max(temp, Math.max(num*min, num));
            min = Math.min(temp, Math.min(num*min, num));
            res = Math.max(res, max);
            System.out.println(max + " "+ min+ " "+ res);
        }
        return res;
    }
}
